src/openms/source/CHEMISTRY/TheoreticalSpectrumGeneratorXLMS.cpp

```diff
@@ -0,0 +1,518 @@
+
+    // precursor intensity
+    defaults_.setValue("precursor_intensity", 1.0, "Intensity of the precursor peak");
+    defaults_.setValue("precursor_H2O_intensity", 1.0, "Intensity of the H2O loss peak of the precursor");
+    defaults_.setValue("precursor_NH3_intensity", 1.0, "Intensity of the NH3 loss peak of the precursor");
+
+    defaultsToParam_();
+  }
+
+  TheoreticalSpectrumGeneratorXLMS::TheoreticalSpectrumGeneratorXLMS(const TheoreticalSpectrumGeneratorXLMS & rhs) :
+    DefaultParamHandler(rhs)
+  {
+  }
+
+  TheoreticalSpectrumGeneratorXLMS & TheoreticalSpectrumGeneratorXLMS::operator=(const TheoreticalSpectrumGeneratorXLMS & rhs)
+  {
+    if (this != &rhs)
+    {
+      DefaultParamHandler::operator=(rhs);
+    }
+    return *this;
+  }
+
+  TheoreticalSpectrumGeneratorXLMS::~TheoreticalSpectrumGeneratorXLMS()
+  {
+  }
+
+  void TheoreticalSpectrumGeneratorXLMS::getCommonIonSpectrum(PeakSpectrum & spectrum, AASequence peptide, Size link_pos, bool frag_alpha, int charge, Size link_pos_2) const
+  {
+    PeakSpectrum::IntegerDataArray integer_array;
+    PeakSpectrum::StringDataArray string_array;
+
+    integer_array.setName("charge");
+    string_array.setName("IonName");
+
+    for (Int z = 1; z <= charge; ++z)
+    {
+      if (add_b_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::BIon, z, link_pos_2);
+      }
+      if (add_y_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::YIon, z, link_pos_2);
+      }
+      if (add_a_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::AIon, z, link_pos_2);
+      }
+      if (add_x_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::XIon, z, link_pos_2);
+      }
+      if (add_c_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::CIon, z, link_pos_2);
+      }
+      if (add_z_ions_)
+      {
+        addCommonPeaks(spectrum, integer_array, string_array, peptide, link_pos, frag_alpha, Residue::ZIon, z, link_pos_2);
+      }
+    }
+
+    spectrum.getIntegerDataArrays().push_back(integer_array);
+    spectrum.getStringDataArrays().push_back(string_array);
+    spectrum.sortByPosition();
+    return;
+  }
+
+void TheoreticalSpectrumGeneratorXLMS::addCommonPeaks(PeakSpectrum & spectrum, PeakSpectrum::IntegerDataArray & integer_array, PeakSpectrum::StringDataArray & string_array, AASequence peptide, Size link_pos, bool frag_alpha, Residue::ResidueType res_type, int charge, Size link_pos_2) const
+  {
+    if (peptide.empty())
+    {
+      cout << "Warning: Attempt at creating XLink Ions Spectrum from empty string!" << endl;
+      return;
+    }
+
+    String ion_type;
```

Comment:
you will need to document `ion_type` and explain what alpha/beta is (see above in the header file)